File: dsa in python/binary_search.py
# iternative 
# def binary_serach(nums,target):
#     nums.sort()
#     n=len(nums)
#     low=0
#     high=n-1
#     while low <= high :
#         mid = (low+high)//2
#         if nums[mid]==target:
#             return mid
#         if nums[mid]<target:
#             low+=1
#         else:
#             high-=1
# nums=[1,2,3,4,5,6,7,8,9]
# print(binary_serach(nums,9))

#recursive solution
def binary_search(nums,low,high):
    target=5
    if low >high:
        return -1
    mid=(low+high)//2
    if nums[mid]==target:
        return mid
    if nums[mid] < target:
        return binary_search(nums,mid+1,high)
    else:
        return binary_search(nums,low,mid-1)
nums=[1,2,3,4,5,6]
n=len(nums)
print(binary_search(nums,0,n-1))
# tc == o(logn)
# sc == o(1)

File: dsa in c/binary_search.c
#include <stdio.h>
#include <stdlib.h>
int binary_search(int arr[], int n, int key)
{
    int low = 0;
    int high = n - 1;
    while (low <= high)
    {
        int mid = (low + high) / 2;
        if (arr[mid] == key)
        {
            return mid;
        }
        else if (arr[mid] > key)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }
    return -1;
}

int main()
{
    int arr[] = {1, 2, 3, 4, 5};
    int n = sizeof(arr) / sizeof(arr[0]);
    int x = binary_search(arr, n, 3);
    if (x == -1)
    {
        printf("Element not found");
    }
    else
    {
        printf("%d", x);
    }
    return 0;
}